File: go/analyze/rules_crossbundle.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// SchemaObjectSet is a simplified representation of the live schema
// state, used for cross-bundle breaking change detection. Populated
// from drift.Snapshot by the reconciler before calling the analyzer.
type SchemaObjectSet struct {
	// Tables maps table name → set of column names.
	Tables map[string]map[string]bool

	// Indexes is the set of index names that exist in the schema.
	Indexes map[string]bool

	// Constraints is the set of constraint names that exist.
	Constraints map[string]bool
}

// CrossBundleBreakDetector detects destructive operations in the
// current bundle that would break objects depended on by other pending
// bundles or by the live application. Unlike the bundle-local
// BreakingChangeDetector, this analyzer has visibility into:
//
//  1. The live schema state (SchemaObjects) — what currently exists
//  2. Other pending bundles' SQL (PendingBundleSQL) — what other
//     bundles expect to exist
//
// It flags cases where:
//   - This bundle drops a column that a pending bundle's SQL references
//   - This bundle drops a table that a pending bundle's SQL references
//   - This bundle renames an index that a pending bundle references
//   - This bundle drops an FK target table
//
// Severity is error — cross-bundle breaks cause silent failures in
// the pending bundle's execution.
type CrossBundleBreakDetector struct{}

func (CrossBundleBreakDetector) ID() string { return "cross-bundle-breaking-change" }

func (CrossBundleBreakDetector) Description() string {
	return "detect destructive operations that would break other pending bundles"
}

func (a *CrossBundleBreakDetector) Check(_ context.Context, m *Migration) ([]Finding, error) {
	if m.SchemaObjects == nil || len(m.PendingBundleSQL) == 0 {
		return nil, nil
	}

	var findings []Finding

	// Build the set of objects referenced by pending bundles.
	pendingRefs := extractReferences(m.PendingBundleSQL)

	// Scan this bundle for destructive operations against those references.
	for _, f := range m.Files {
		findings = append(findings, checkDropTableCrossBundle(f, pendingRefs, m.SchemaObjects)...)
		findings = append(findings, checkDropColumnCrossBundle(f, pendingRefs, m.SchemaObjects)...)
		findings = append(findings, checkRenameColumnCrossBundle(f, pendingRefs, m.SchemaObjects)...)
		findings = append(findings, checkDropIndexCrossBundle(f, pendingRefs)...)
	}

	return findings, nil
}

// pendingReferences captures what objects pending bundles' SQL mentions.
type pendingReferences struct {
	// TableRefs maps table names referenced in pending SQL.
	TableRefs map[string]bool
	// ColumnRefs maps "table.column" to true for columns referenced.
	ColumnRefs map[string]bool
	// IndexRefs maps index names referenced.
	IndexRefs map[string]bool
}

var (
	// Match table names in common SQL contexts.
	reTableRef = regexp.MustCompile(
		`(?i)\b(?:FROM|JOIN|INTO|UPDATE|TABLE|REFERENCES)\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)`)
	// Match column references in ALTER TABLE context.
	reColumnRef = regexp.MustCompile(
		`(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+(?:ADD|ALTER|DROP)\s+(?:COLUMN\s+)?(?:IF\s+(?:NOT\s+)?EXISTS\s+)?([a-z_][a-z0-9_]*)`)
	// Match index references in CREATE/DROP INDEX.
	reIndexRef = regexp.MustCompile(
		`(?i)\b(?:CREATE|DROP)\s+(?:UNIQUE\s+)?INDEX\s+(?:CONCURRENTLY\s+)?(?:IF\s+(?:NOT\s+)?EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)`)
	// Cross-bundle regex for DROP TABLE.
	reCBDropTable = regexp.MustCompile(
		`(?i)\bDROP\s+TABLE\s+(?:IF\s+EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)`)
	// Cross-bundle regex for DROP COLUMN.
	reCBDropColumn = regexp.MustCompile(
		`(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+DROP\s+(?:COLUMN\s+)?(?:IF\s+EXISTS\s+)?([a-z_][a-z0-9_]*)`)
	// Cross-bundle regex for RENAME COLUMN.
	reCBRenameColumn = regexp.MustCompile(
		`(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+RENAME\s+(?:COLUMN\s+)?([a-z_][a-z0-9_]*)\s+TO\s+([a-z_][a-z0-9_]*)`)
	// Cross-bundle regex for DROP INDEX.
	reCBDropIndex = regexp.MustCompile(
		`(?i)\bDROP\s+INDEX\s+(?:CONCURRENTLY\s+)?(?:IF\s+EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)`)
)

// extractReferences scans pending bundle SQL to find referenced objects.
func extractReferences(pendingSQL []FileBody) *pendingReferences {
	refs := &pendingReferences{
		TableRefs:  make(map[string]bool),
		ColumnRefs: make(map[string]bool),
		IndexRefs:  make(map[string]bool),
	}

	for _, f := range pendingSQL {
		for _, match := range reTableRef.FindAllStringSubmatch(f.Body, -1) {
			refs.TableRefs[strings.ToLower(match[1])] = true
		}
		for _, match := range reColumnRef.FindAllStringSubmatch(f.Body, -1) {
			table := strings.ToLower(match[1])
			col := strings.ToLower(match[2])
			refs.ColumnRefs[table+"."+col] = true
			refs.TableRefs[table] = true
		}
		for _, match := range reIndexRef.FindAllStringSubmatch(f.Body, -1) {
			refs.IndexRefs[strings.ToLower(match[1])] = true
		}
	}

	return refs
}

func checkDropTableCrossBundle(f FileBody, refs *pendingReferences, _ *SchemaObjectSet) []Finding {
	var findings []Finding
	for _, match := range reCBDropTable.FindAllStringSubmatchIndex(f.Body, -1) {
		table := strings.ToLower(f.Body[match[2]:match[3]])
		if refs.TableRefs[table] {
			line := lineOfOffset(f.Body, match[0])
			findings = append(findings, Finding{
				Rule:     "cross-bundle-breaking-change",
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(line),
				Message: fmt.Sprintf(
					"DROP TABLE %s would break a pending bundle that references this table",
					table),
			})
		}
	}
	return findings
}

func checkDropColumnCrossBundle(f FileBody, refs *pendingReferences, _ *SchemaObjectSet) []Finding {
	var findings []Finding
	for _, match := range reCBDropColumn.FindAllStringSubmatchIndex(f.Body, -1) {
		table := strings.ToLower(f.Body[match[2]:match[3]])
		col := strings.ToLower(f.Body[match[4]:match[5]])
		key := table + "." + col
		if refs.ColumnRefs[key] {
			line := lineOfOffset(f.Body, match[0])
			findings = append(findings, Finding{
				Rule:     "cross-bundle-breaking-change",
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(line),
				Message: fmt.Sprintf(
					"DROP COLUMN %s.%s would break a pending bundle that references this column",
					table, col),
			})
		}
	}
	return findings
}

func checkRenameColumnCrossBundle(f FileBody, refs *pendingReferences, _ *SchemaObjectSet) []Finding {
	var findings []Finding
	for _, match := range reCBRenameColumn.FindAllStringSubmatchIndex(f.Body, -1) {
		table := strings.ToLower(f.Body[match[2]:match[3]])
		oldCol := strings.ToLower(f.Body[match[4]:match[5]])
		key := table + "." + oldCol
		if refs.ColumnRefs[key] {
			newCol := strings.ToLower(f.Body[match[6]:match[7]])
			line := lineOfOffset(f.Body, match[0])
			findings = append(findings, Finding{
				Rule:     "cross-bundle-breaking-change",
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(line),
				Message: fmt.Sprintf(
					"RENAME COLUMN %s.%s TO %s would break a pending bundle that references the old name",
					table, oldCol, newCol),
			})
		}
	}
	return findings
}

func checkDropIndexCrossBundle(f FileBody, refs *pendingReferences) []Finding {
	var findings []Finding
	for _, match := range reCBDropIndex.FindAllStringSubmatchIndex(f.Body, -1) {
		idx := strings.ToLower(f.Body[match[2]:match[3]])
		if refs.IndexRefs[idx] {
			line := lineOfOffset(f.Body, match[0])
			findings = append(findings, Finding{
				Rule:     "cross-bundle-breaking-change",
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(line),
				Message: fmt.Sprintf(
					"DROP INDEX %s would break a pending bundle that references this index",
					idx),
			})
		}
	}
	return findings
}

File: rust/keystone-sdk/src/analyze/rules_crossbundle.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase 9.6 — cross-bundle breaking change detection. Ported from
//! `rules_crossbundle.go`. Flags destructive ops in this bundle that would
//! break objects referenced by other pending bundles' SQL.

use std::collections::{HashMap, HashSet};
use std::sync::LazyLock;

use regex::Regex;

use super::{finding, line_of_offset, Analyzer, FileBody, Finding, LintLevel, Migration};

/// Simplified live-schema state, populated from `drift::Snapshot` by the
/// reconciler. (The cross-bundle rule only checks for its *presence*; the
/// fields mirror the Go type for API completeness.)
#[derive(Debug, Clone, Default)]
pub struct SchemaObjectSet {
    /// Table name → set of column names.
    pub tables: HashMap<String, HashSet<String>>,
    /// Index names that exist in the schema.
    pub indexes: HashSet<String>,
    /// Constraint names that exist.
    pub constraints: HashSet<String>,
}

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(
    RE_TABLE_REF,
    r"(?i)\b(?:FROM|JOIN|INTO|UPDATE|TABLE|REFERENCES)\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)"
);
re!(
    RE_COLUMN_REF,
    r"(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+(?:ADD|ALTER|DROP)\s+(?:COLUMN\s+)?(?:IF\s+(?:NOT\s+)?EXISTS\s+)?([a-z_][a-z0-9_]*)"
);
re!(
    RE_INDEX_REF,
    r"(?i)\b(?:CREATE|DROP)\s+(?:UNIQUE\s+)?INDEX\s+(?:CONCURRENTLY\s+)?(?:IF\s+(?:NOT\s+)?EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)"
);
re!(
    RE_CB_DROP_TABLE,
    r"(?i)\bDROP\s+TABLE\s+(?:IF\s+EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)"
);
re!(
    RE_CB_DROP_COLUMN,
    r"(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+DROP\s+(?:COLUMN\s+)?(?:IF\s+EXISTS\s+)?([a-z_][a-z0-9_]*)"
);
re!(
    RE_CB_RENAME_COLUMN,
    r"(?i)\bALTER\s+TABLE\s+(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)\s+RENAME\s+(?:COLUMN\s+)?([a-z_][a-z0-9_]*)\s+TO\s+([a-z_][a-z0-9_]*)"
);
re!(
    RE_CB_DROP_INDEX,
    r"(?i)\bDROP\s+INDEX\s+(?:CONCURRENTLY\s+)?(?:IF\s+EXISTS\s+)?(?:(?:[a-z_][a-z0-9_]*)\.)?([a-z_][a-z0-9_]*)"
);

const RULE: &str = "cross-bundle-breaking-change";

/// cross-bundle-breaking-change
pub struct CrossBundleBreakDetector;
impl Analyzer for CrossBundleBreakDetector {
    fn id(&self) -> &'static str {
        RULE
    }
    fn description(&self) -> &'static str {
        "detect destructive operations that would break other pending bundles"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        if m.schema_objects.is_none() || m.pending_bundle_sql.is_empty() {
            return Vec::new();
        }
        let refs = extract_references(&m.pending_bundle_sql);
        let mut findings = Vec::new();
        for f in &m.files {
            findings.extend(check_drop_table(f, &refs));
            findings.extend(check_drop_column(f, &refs));
            findings.extend(check_rename_column(f, &refs));
            findings.extend(check_drop_index(f, &refs));
        }
        findings
    }
}

/// What objects pending bundles' SQL references.
#[derive(Default)]
struct PendingReferences {
    table_refs: HashSet<String>,
    /// "table.column" keys.
    column_refs: HashSet<String>,
    index_refs: HashSet<String>,
}

fn extract_references(pending_sql: &[FileBody]) -> PendingReferences {
    let mut refs = PendingReferences::default();
    for f in pending_sql {
        for caps in RE_TABLE_REF.captures_iter(&f.body) {
            refs.table_refs
                .insert(caps.get(1).unwrap().as_str().to_lowercase());
        }
        for caps in RE_COLUMN_REF.captures_iter(&f.body) {
            let table = caps.get(1).unwrap().as_str().to_lowercase();
            let col = caps.get(2).unwrap().as_str().to_lowercase();
            refs.column_refs.insert(format!("{table}.{col}"));
            refs.table_refs.insert(table);
        }
        for caps in RE_INDEX_REF.captures_iter(&f.body) {
            refs.index_refs
                .insert(caps.get(1).unwrap().as_str().to_lowercase());
        }
    }
    refs
}

fn check_drop_table(f: &FileBody, refs: &PendingReferences) -> Vec<Finding> {
    let mut out = Vec::new();
    for caps in RE_CB_DROP_TABLE.captures_iter(&f.body) {
        let whole = caps.get(0).unwrap().start();
        let table = caps.get(1).unwrap().as_str().to_lowercase();
        if refs.table_refs.contains(&table) {
            out.push(finding(
                RULE,
                LintLevel::Error,
                &f.name,
                line_of_offset(&f.body, whole),
                format!(
                    "DROP TABLE {table} would break a pending bundle that references this table"
                ),
            ));
        }
    }
    out
}

fn check_drop_column(f: &FileBody, refs: &PendingReferences) -> Vec<Finding> {
    let mut out = Vec::new();
    for caps in RE_CB_DROP_COLUMN.captures_iter(&f.body) {
        let whole = caps.get(0).unwrap().start();
        let table = caps.get(1).unwrap().as_str().to_lowercase();
        let col = caps.get(2).unwrap().as_str().to_lowercase();
        if refs.column_refs.contains(&format!("{table}.{col}")) {
            out.push(finding(
                RULE,
                LintLevel::Error,
                &f.name,
                line_of_offset(&f.body, whole),
                format!(
                    "DROP COLUMN {table}.{col} would break a pending bundle that references this column"
                ),
            ));
        }
    }
    out
}

fn check_rename_column(f: &FileBody, refs: &PendingReferences) -> Vec<Finding> {
    let mut out = Vec::new();
    for caps in RE_CB_RENAME_COLUMN.captures_iter(&f.body) {
        let whole = caps.get(0).unwrap().start();
        let table = caps.get(1).unwrap().as_str().to_lowercase();
        let old_col = caps.get(2).unwrap().as_str().to_lowercase();
        if refs.column_refs.contains(&format!("{table}.{old_col}")) {
            let new_col = caps.get(3).unwrap().as_str().to_lowercase();
            out.push(finding(
                RULE,
                LintLevel::Error,
                &f.name,
                line_of_offset(&f.body, whole),
                format!(
                    "RENAME COLUMN {table}.{old_col} TO {new_col} would break a pending bundle that references the old name"
                ),
            ));
        }
    }
    out
}

fn check_drop_index(f: &FileBody, refs: &PendingReferences) -> Vec<Finding> {
    let mut out = Vec::new();
    for caps in RE_CB_DROP_INDEX.captures_iter(&f.body) {
        let whole = caps.get(0).unwrap().start();
        let idx = caps.get(1).unwrap().as_str().to_lowercase();
        if refs.index_refs.contains(&idx) {
            out.push(finding(
                RULE,
                LintLevel::Error,
                &f.name,
                line_of_offset(&f.body, whole),
                format!("DROP INDEX {idx} would break a pending bundle that references this index"),
            ));
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    #[test]
    fn no_schema_objects_means_silent() {
        let m = Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: "DROP TABLE t;".into(),
            }],
            pending_bundle_sql: vec![FileBody {
                name: "p.sql".into(),
                body: "SELECT * FROM t;".into(),
            }],
            schema_objects: None,
            ..Default::default()
        };
        assert_eq!(CrossBundleBreakDetector.check(&m).len(), 0);
    }

    #[test]
    fn drop_table_referenced_by_pending() {
        let m = Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: "DROP TABLE orders;".into(),
            }],
            pending_bundle_sql: vec![FileBody {
                name: "p.sql".into(),
                body: "SELECT * FROM orders;".into(),
            }],
            schema_objects: Some(SchemaObjectSet::default()),
            ..Default::default()
        };
        let f = CrossBundleBreakDetector.check(&m);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].rule, "cross-bundle-breaking-change");
        assert!(f[0].message.contains("DROP TABLE orders"));
    }

    #[test]
    fn drop_column_referenced_by_pending() {
        let m = Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: "ALTER TABLE orders DROP COLUMN total;".into(),
            }],
            pending_bundle_sql: vec![FileBody {
                name: "p.sql".into(),
                body: "ALTER TABLE orders ALTER COLUMN total TYPE bigint;".into(),
            }],
            schema_objects: Some(SchemaObjectSet::default()),
            ..Default::default()
        };
        let f = CrossBundleBreakDetector.check(&m);
        assert_eq!(f.len(), 1);
        assert!(f[0].message.contains("DROP COLUMN orders.total"));
    }
}
